sklearn/ensemble/bagging.py

```diff
@@ -310,8 +310,12 @@ def _fit(self, X, y, max_samples=None, max_depth=None, sample_weight=None):
         # Validate max_features
         if isinstance(self.max_features, (numbers.Integral, np.integer)):
             max_features = self.max_features
-        else:  # float
-            max_features = int(self.max_features * self.n_features_)
+        elif isinstance(self.max_features, (numbers.Real, np.float)):
+            if not self.max_features > 0.0:
+                raise ValueError("max_features must be in (0, n_features]")
+            max_features = max(1, int(self.max_features * self.n_features_))
+        else:
+            raise ValueError("max_features must be int or float")
 
         if not (0 < max_features <= self.n_features_):
```

Comment:
Put this before the above test and you can simplify the code. Don't worry about validating that it is numeric. Comparing to a number is good enough for that unexpected case.

Reply:
Thanks for the review @jnothman ! I'm struggling to find a simpler implementation that still passes other tests. Could you expand on your comment above?

The desired behaviour is to raise a Value error if `self.max_features` is 0 or "foobar", but not raise an error if it is 0.1. Relevant existing unit tests: https://github.com/scikit-learn/scikit-learn/blob/fa4de831a0eca4d1f369b972b0900b23e1348b4f/sklearn/ensemble/tests/test_bagging.py#L424-L434

Comparing self.max_features to a number without first casting to an integer causes an unexpected Type error if max_features is a string: `TypeError: unorderable types: int() < str()`.

However, casting to an integer means that 0.1 would be rounded down to 0, and hence a Value error is raised (which is the behaviour the PR is trying to avoid).

Reply:
Well... TypeError might really be the more appropriate error anyway, but let's not quibble with the tests. Why not:

```
if not numeric:
   raise ValueError
if real:
   max_features = max_features * features
if not 0 < max_features <= n_features:
   raise ValueError
max_features = int(max_features)
```
but perhaps that logic is no less complicated than the present?

Reply:
Good point: I could just modify that test to expect a TypeError if a string is input rather than a ValueError.

Here's a suggested modification of the logic that ensures 0.1 is not rounded down to zero:
```
if isinstance(self.max_features, (numbers.Integral, np.integer)):
    max_features = self.max_features
else:  # float
   max_features = self.max_features * self.n_features_

if not (0 < max_features <= self.n_features_):
    raise ValueError

max_features = max(1, int(max_features))
```